libs/cardano-ledger-binary/src/Cardano/Ledger/Binary/Decoding/Annotated.hs

```diff
@@ -241,6 +246,32 @@ decodeAnnSet dec = do
     pure s
 {-# INLINE decodeAnnSet #-}
 
+decodeNonEmptySetLikeEnforceNoDuplicatesAnn ::
+  forall s a b c.
+  (Monoid b, DecCBOR (Annotator a)) =>
+  -- | Add an element into the decoded Set like data structure
+  (a -> b -> b) ->
+  -- | Get the final data structure and the number of elements it has.
+  (b -> (Int, c)) ->
+  Decoder s (Annotator c)
+decodeNonEmptySetLikeEnforceNoDuplicatesAnn insert getFinalWithLen = do
+  allowTag setTag
+  valAnns <- decodeNonEmptyList decCBOR
+  pure $ go (mempty, 0) valAnns
+  where
+    go :: (b, Int) -> NonEmpty (Annotator a) -> Annotator c
+    go (!m, !n) (x :| xs) = do
+      val <- x
+      case xs of
+        [] -> finish (insert val m) (n + 1)
+        (y : ys) -> go (insert val m, n + 1) (y :| ys)
+    finish :: b -> Int -> Annotator c
+    finish m n
+      | finalLen /= n = fail "Duplicates found, expected no duplicates"
+      | otherwise = pure final
+      where
+        (finalLen, final) = getFinalWithLen m
```

Comment:
We should add some inline pragmas, since decoders do benefit from it quite a bit
```suggestion
        (finalLen, final) = getFinalWithLen m
    {-# INLINE finish #-}
{-# INLINE decodeNonEmptySetLikeEnforceNoDuplicatesAnn #-}
```